### DISCUSSION TITLE
easeBy action goes through colliders

### DISCUSSION THREAD

**Discussion by ikiselev1989** [excaliburjs/Excalibur#2802] [Q&A]
There is seem like issue in physic engine. I expect that active collider will collide with fixed one and will freezed but it's going through.
[code](https://stackblitz.com/edit/typescript-shetwa?file=index.ts)

```
import {
  Engine,
  CollisionGroupManager,
  Actor,
  CollisionGroup,
  CollisionType,
  vec,
  EasingFunctions,
  Physics,
} from 'excalibur';

const engine = new Engine({
  width: 600,
  height: 400,
});

engine.start();

engine.showDebug(true);

const player = new Actor({
  width: 50,
  height: 50,
  collisionType: CollisionType.Active,
});

engine.add(player);

const wall = new Actor({
  width: 50,
  height: 200,
  pos: vec(200, 0),
  collisionType: CollisionType.Fixed,
});

engine.add(wall);

engine.input.keyboard.on('release', () => {
  player.pos.setTo(0, 0);
  player.actions.easeBy(vec(400, 0), 500, EasingFunctions.EaseInOutQuad);
});
```

**Comment by eonarheim** [comment]
Hi @ikiselev1989 

Good find! I think this has something to do with the implementation of the easing action. The easing implementation progressively increases the velocity if the position isn't where it should be according to the easing function, which causing the actor to update over the collider eventually when the velocity is high enough! I'll see if there is a way to modify the implementation of easing to play nice with collisions.

EaseTo: https://github.com/excaliburjs/Excalibur/blob/main/src/engine/Actions/Action/EaseTo.ts#L60
EaseBy: https://github.com/excaliburjs/Excalibur/blob/main/src/engine/Actions/Action/EaseBy.ts#L62

Here is the bug to track: https://github.com/excaliburjs/Excalibur/issues/2803

If you switch to `player.actions.moveBy(...)` or `player.actions.moveTo(...)` things seem to work as you expect and collide. That's because the moveBy/moveTo modifies the velocity by a constant amount.

**Reply by eonarheim** [reply]
The update problem, becomes more clear if you set the duration to be much longer in `easeBy`

**Reply by eonarheim** [reply]
Confirmed, the console log output shows a series of 0's for velocity when colliding starts, then a BIG velocity that causes the actor to update through the wall.

```typescript
let currentlyEasing = false;
const player = new Actor({
  width: 50,
  height: 50,
  collisionType: CollisionType.Active,
});
player.onPostUpdate = () => {
  if (currentlyEasing) {
    console.log(player.vel.x);
  }
}

engine.add(player);

const wall = new Actor({
  width: 50,
  height: 200,
  pos: vec(200, 0),
  collisionType: CollisionType.Fixed,
});

engine.add(wall);

engine.input.keyboard.on('release', async () => {
  player.pos.setTo(0, 0);
  currentlyEasing = true;
  await player.actions.easeBy(vec(400, 0), 3500, EasingFunctions.EaseInOutQuad).toPromise();
  currentlyEasing = false;
});

```

**Reply by eonarheim** [reply]
@ikiselev1989 I have a somewhat unideal workaround. This will suspend all actions if the collision direction is in the same direction roughly as the velocity.

```typescript
player.on('precollision', (evt) => {
  // if collision is mostly in the direction of intersection suspend all actions
  const dir = Math.abs(
    evt.intersection.normalize().dot(player.vel.normalize())
  );
  console.log(dir);
  if (dir > 0.5) {
    player.actions.clearActions();
  }
});

```

**Reply by ikiselev1989** [reply]
@eonarheim thanks for your feedback!
I tried to use `moveBy` and this is working the same if I set wall's width to 25 and moving speed to 5000 :)
```
const wall = new Actor({
  width: 25,
  height: 200,
  pos: vec(200, 0),
  collisionType: CollisionType.Fixed,
});

engine.add(wall);

engine.input.keyboard.on('release', () => {
  player.pos.setTo(0, 0);
  player.actions.moveBy(vec(400, 0), 5000);
});
```

**Reply by ikiselev1989** [reply]
So I will use your unideal workaround :) thanks!

**Reply by eonarheim** [reply]
I'll write this in the bug, but I think the most robust long term fix will be to improve the continuous collision solver code. That way it won't matter how fast objects are traveling, they'll be a shape cast or ray cast to prevent tunneling

**Reply by itsscain** [reply]
@ikiselev1989 
```
import {
  Engine,
  Actor,
  CollisionType,
  vec,
  EasingFunctions,
} from 'excalibur';

const engine = new Engine({
  width: 600,
  height: 400,
});

engine.start();
engine.showDebug(true);

const player = new Actor({
  width: 50,
  height: 50,
  collisionType: CollisionType.Active,
});

// Collision event listener to stop actions when collision is detected
player.on('precollision', (evt) => {
  const dir = Math.abs(evt.intersection.normalize().dot(player.vel.normalize()));
  if (dir > 0.5) {
    player.actions.clearActions();
  }
});

engine.add(player);

const wall = new Actor({
  width: 50,
  height: 200,
  pos: vec(200, 0),
  collisionType: CollisionType.Fixed,
});

engine.add(wall);

engine.input.keyboard.on('release', () => {
  player.pos.setTo(0, 0);
  player.actions.easeBy(vec(400, 0), 3500, EasingFunctions.EaseInOutQuad);
});
```

use `player.on('precollision', (evt) => {...})` to detect collisions and clear actions if the collision direction is roughly in the same direction as the player's velocity.
The `easeBy` action is still used, but the collision listener should prevent the player from moving through the wall.
